tests/tests/test_filetransfer.py

```diff
@@ -281,4 +323,406 @@ def test_filetransfer(self, enterprise_no_client):
+        assert r.status_code == 201
+
+        set_limits(
+            mender_device,
+            {
+                "Enabled": True,
+                "FileTransfer": {
+                    "Chroot": "/var/lib/mender/filetransfer",
+                    "FollowSymLinks": True,  # in the image /var/lib/mender is a symlink
+                    "PreserveMode": True,
+                },
+            },
+        )
+        time.sleep(4)  # wait for mender-connect re-connect
+
+        logger.info("-- testcase: File Transfer limits: preserve modes;")
+        f = NamedTemporaryFile(delete=False)
+        f.write(os.urandom(16))
+        f.close()
+        r = upload_file(
+            "/var/lib/mender/filetransfer/modes.bin",
+            open(f.name, "rb"),
+            devid,
+            authtoken,
+            mode="4711",
+        )
+        modes_ls = mender_device.run("ls -al /var/lib/mender/filetransfer/modes.bin")
+        logger.info(
+            "test_filetransfer_limits_upload ls -al /var/lib/mender/filetransfer/modes.bin:\n%s"
+            % modes_ls
+        )
+
+        assert modes_ls.startswith("-rws--x--x")
+        assert r.status_code == 201
+
+        set_limits(
+            mender_device,
+            {
+                "Enabled": True,
+                "FileTransfer": {
+                    "Chroot": "/var/lib/mender/filetransfer",
+                    "FollowSymLinks": True,  # in the image /var/lib/mender is a symlink
+                    "PreserveOwner": True,
+                    "PreserveGroup": True,
+                },
+            },
+        )
+        time.sleep(4)  # wait for mender-connect re-connect
+
+        logger.info("-- testcase: File Transfer limits: preserve owner and group;")
+        f = NamedTemporaryFile(delete=False)
+        f.write(os.urandom(16))
+        f.close()
+        gid = int(mender_device.run("cat /etc/group | tail -1| cut -f3 -d:"))
+        uid = int(mender_device.run("cat /etc/passwd | tail -1| cut -f3 -d:"))
+        logger.info("test_filetransfer_limits_upload gid/uid %d/%d", gid, uid)
+        r = upload_file(
+            "/var/lib/mender/filetransfer/ownergroup.bin",
+            open(f.name, "rb"),
+            devid,
+            authtoken,
+            uid=str(uid),
+            gid=str(gid),
+        )
+        owner_group = mender_device.run(
+            "ls -aln /var/lib/mender/filetransfer/ownergroup.bin | cut -f 3,4 -d' '"
+        )
+
+        assert owner_group == str(uid) + " " + str(gid) + "\n"
+        assert r.status_code == 201
+
+    def test_filetransfer_limits_download(self, standard_setup_one_client):
+        mender_device, devid, authtoken = self.test_filetransfer_limits_setup(
+            standard_setup_one_client
+        )
+        """Tests the file transfer features with limits"""
+        set_limits(
+            mender_device,
+            {
+                "Enabled": True,
+                "FileTransfer": {"Chroot": "/var/lib/mender/filetransfer",},
```

Comment:
Can you remove all the commas before the closed parenthesis? This is because of black, I believe.
Simply replace all `.}` with `}`.

Reply:
right. due to auto formatter. fixed.